File: 3-SAT/FormulaCreator/JoinMethods.cs
using System.Collections.Generic;
using System.Linq;

namespace FormulaCreator
{
    public static class JoinMethods
    {
        public static IEnumerable<IJoinMethod> GetMethods()
        {
            return new IJoinMethod[]
                       {
                           new Join3AsIs(),
                           new JoinBetweenTiers2(),
                           new JoinLeft2(),
                           new JoinRight2(),
                           new JoinLeft1(),
                           new JoinRight1(),
                           new JoinRight0(),
                       };
        }
    }

    public class JoinInfo
    {
        public int ConcatenationPower { get; set; }
        public ITriplet TargetPremutation { get; set; }
        public IJoinMethod JoinMethod { get; private set; }
        public string Rule { get; set; }

        public JoinInfo(IJoinMethod method)
        {
            JoinMethod = method;
            ConcatenationPower = -1;
        }
    }

    public interface IJoinMethod
    {
        JoinInfo GetJoinInfo(ITabularFormula formula, ITriplet triplet);
    }

    internal class Join3AsIs : IJoinMethod
    {
        public JoinInfo GetJoinInfo(ITabularFormula formula, ITriplet triplet)
        {
            var result = new JoinInfo(this);

            var aIndex = formula.Premutation.IndexOf(triplet.AName);
            var bIndex = formula.Premutation.IndexOf(triplet.BName);
            var cIndex = formula.Premutation.IndexOf(triplet.CName);

            if (aIndex >= 0 && bIndex >= 0 && cIndex >= 0)
            {
                if (Follows(aIndex, bIndex, cIndex))
                {
                    result.Rule = "1";
                    result.ConcatenationPower = 3;
                    result.TargetPremutation = triplet;
                }
                else if (Follows(bIndex, aIndex, cIndex))
                {
                    result.Rule = "2";
                    result.ConcatenationPower = 3;
                    result.TargetPremutation = new SimpleTriplet(triplet.BName, triplet.AName, triplet.CName);
                }
                else if (Follows(aIndex, cIndex, bIndex))
                {
                    result.Rule = "3";
                    result.ConcatenationPower = 3;
                    result.TargetPremutation = new SimpleTriplet(triplet.AName, triplet.CName, triplet.BName);
                }
                else if (Follows(cIndex, aIndex, bIndex))
                {
                    result.Rule = "4";
                    result.ConcatenationPower = 3;
                    result.TargetPremutation = new SimpleTriplet(triplet.CName, triplet.AName, triplet.BName);
                }
                else if (Follows(bIndex, cIndex, aIndex))
                {
                    result.Rule = "5";
                    result.ConcatenationPower = 3;
                    result.TargetPremutation = new SimpleTriplet(triplet.BName, triplet.CName, triplet.AName);
                }
                else if (Follows(cIndex, bIndex, aIndex))
                {
                    result.Rule = "6";
                    result.ConcatenationPower = 3;
                    result.TargetPremutation = new SimpleTriplet(triplet.CName, triplet.BName, triplet.AName);
                }
            }

            return result;
        }

        private static bool Follows(int aIndex, int bIndex, int cIndex)
        {
            return Follows(aIndex, bIndex) && Follows(bIndex, cIndex);
        }

        private static bool Follows(int aIndex, int bIndex)
        {
            return aIndex == bIndex - 1;
        }
    }

    internal class JoinLeft1 : IJoinMethod
    {
        public JoinInfo GetJoinInfo(ITabularFormula formula, ITriplet triplet)
        {
            var result = new JoinInfo(this);

            if (formula.Premutation[0] == triplet.CName
                && !formula.Premutation.Contains(triplet.BName)
                && !formula.Premutation.Contains(triplet.AName))
            {
                result.Rule = "1";
                result.ConcatenationPower = 1;
                result.TargetPremutation = new SimpleTriplet(triplet.AName, triplet.BName, triplet.CName);
            }
            else if (formula.Premutation[0] == triplet.BName
                && !formula.Premutation.Contains(triplet.AName)
                && !formula.Premutation.Contains(triplet.CName))
            {
                result.Rule = "2";
                result.ConcatenationPower = 1;
                result.TargetPremutation = new SimpleTriplet(triplet.AName, triplet.CName, triplet.BName);
            }
            else if (formula.Premutation[0] == triplet.AName
                && !formula.Premutation.Contains(triplet.BName)
                && !formula.Premutation.Contains(triplet.CName))
            {
                result.Rule = "3";
                result.ConcatenationPower = 1;
                result.TargetPremutation = new SimpleTriplet(triplet.BName, triplet.CName, triplet.AName);
            }

            return result;
        }
    }

    internal class JoinLeft2 : IJoinMethod
    {
        public JoinInfo GetJoinInfo(ITabularFormula formula, ITriplet triplet)
        {
            var result = new JoinInfo(this);

            if (formula.Premutation[0] == triplet.BName
                && formula.Premutation[1] == triplet.CName
                && !formula.Premutation.Contains(triplet.AName))
            {
                result.Rule = "1";
                result.ConcatenationPower = 2;
                result.TargetPremutation = new SimpleTriplet(triplet.AName, triplet.BName, triplet.CName);
            }
            else if (formula.Premutation[0] == triplet.CName
                && formula.Premutation[1] == triplet.BName
                && !formula.Premutation.Contains(triplet.AName))
            {
                result.Rule = "2";
                result.ConcatenationPower = 2;
                result.TargetPremutation = new SimpleTriplet(triplet.AName, triplet.CName, triplet.BName);
            }
            else if (formula.Premutation[0] == triplet.AName
                && formula.Premutation[1] == triplet.BName
                && !formula.Premutation.Contains(triplet.CName))
            {
                result.Rule = "3";
                result.ConcatenationPower = 2;
                result.TargetPremutation = new SimpleTriplet(triplet.CName, triplet.AName, triplet.BName);
            }
            else if (formula.Premutation[0] == triplet.BName
                && formula.Premutation[1] == triplet.AName
                && !formula.Premutation.Contains(triplet.CName))
            {
                result.Rule = "4";
                result.ConcatenationPower = 2;
                result.TargetPremutation = new SimpleTriplet(triplet.CName, triplet.BName, triplet.AName);
            }
            else if (formula.Premutation[0] == triplet.AName
                && formula.Premutation[1] == triplet.CName
                && !formula.Premutation.Contains(triplet.BName))
            {
                result.Rule = "5";
                result.ConcatenationPower = 2;
                result.TargetPremutation = new SimpleTriplet(triplet.BName, triplet.AName, triplet.CName);
            }
            else if (formula.Premutation[0] == triplet.CName
                && formula.Premutation[1] == triplet.AName
                && !formula.Premutation.Contains(triplet.BName))
            {
                result.Rule = "6";
                result.ConcatenationPower = 2;
                result.TargetPremutation = new SimpleTriplet(triplet.BName, triplet.CName, triplet.AName);
            }

            return result;
        }
    }

    internal class JoinRight2 : IJoinMethod
    {
        public JoinInfo GetJoinInfo(ITabularFormula formula, ITriplet triplet)
        {
            var result = new JoinInfo(this);

            if (formula.Premutation[formula.Premutation.Count - 2] == triplet.AName
                && formula.Premutation[formula.Premutation.Count - 1] == triplet.BName
                && !formula.Premutation.Contains(triplet.CName))
            {
                result.Rule = "1";
                result.ConcatenationPower = 2;
                result.TargetPremutation = triplet;
            }
            else if (formula.Premutation[formula.Premutation.Count - 2] == triplet.BName
                && formula.Premutation[formula.Premutation.Count - 1] == triplet.AName
                && !formula.Premutation.Contains(triplet.CName))
            {
                result.Rule = "2";
                result.ConcatenationPower = 2;
                result.TargetPremutation = new SimpleTriplet(triplet.BName, triplet.AName, triplet.CName);
            }
            else if (formula.Premutation[formula.Premutation.Count - 2] == triplet.BName
                && formula.Premutation[formula.Premutation.Count - 1] == triplet.CName
                && !formula.Premutation.Contains(triplet.AName))
            {
                result.Rule = "3";
                result.ConcatenationPower = 2;
                result.TargetPremutation = new SimpleTriplet(triplet.BName, triplet.CName, triplet.AName);
            }
            else if (formula.Premutation[formula.Premutation.Count - 2] == triplet.CName
                && formula.Premutation[formula.Premutation.Count - 1] == triplet.BName
                && !formula.Premutation.Contains(triplet.AName))
            {
                result.Rule = "4";
                result.ConcatenationPower = 2;
                result.TargetPremutation = new SimpleTriplet(triplet.CName, triplet.BName, triplet.AName);
            }
            else if (formula.Premutation[formula.Premutation.Count - 2] == triplet.AName
                && formula.Premutation[formula.Premutation.Count - 1] == triplet.CName
                && !formula.Premutation.Contains(triplet.BName))
            {
                result.Rule = "5";
                result.ConcatenationPower = 2;
                result.TargetPremutation = new SimpleTriplet(triplet.AName, triplet.CName, triplet.BName);
            }
            else if (formula.Premutation[formula.Premutation.Count - 2] == triplet.CName
                && formula.Premutation[formula.Premutation.Count - 1] == triplet.AName
                && !formula.Premutation.Contains(triplet.BName))
            {
                result.Rule = "6";
                result.ConcatenationPower = 2;
                result.TargetPremutation = new SimpleTriplet(triplet.CName, triplet.AName, triplet.BName);
            }

            return result;
        }
    }

    internal class JoinRight1 : IJoinMethod
    {
        public JoinInfo GetJoinInfo(ITabularFormula formula, ITriplet triplet)
        {
            var result = new JoinInfo(this);

            if (formula.Premutation[formula.Premutation.Count - 1] == triplet.AName
                && !formula.Premutation.Contains(triplet.BName)
                && !formula.Premutation.Contains(triplet.CName))
            {
                result.Rule = "1";
                result.ConcatenationPower = 1;
                result.TargetPremutation = triplet;
            }
            else if (formula.Premutation[formula.Premutation.Count - 1] == triplet.BName
                && !formula.Premutation.Contains(triplet.AName)
                && !formula.Premutation.Contains(triplet.CName))
            {
                result.Rule = "2";
                result.ConcatenationPower = 1;
                result.TargetPremutation = new SimpleTriplet(triplet.BName, triplet.AName, triplet.CName);
            }
            else if (formula.Premutation[formula.Premutation.Count - 1] == triplet.CName
                && !formula.Premutation.Contains(triplet.AName)
                && !formula.Premutation.Contains(triplet.BName))
            {
                result.Rule = "3";
                result.ConcatenationPower = 1;
                result.TargetPremutation = new SimpleTriplet(triplet.CName, triplet.AName, triplet.BName);
            }

            return result;
        }
    }

    internal class JoinRight0 : IJoinMethod
    {
        public JoinInfo GetJoinInfo(ITabularFormula formula, ITriplet triplet)
        {
            var result = new JoinInfo(this);

            if (!formula.Premutation.Contains(triplet.AName)
                && !formula.Premutation.Contains(triplet.BName)
                && !formula.Premutation.Contains(triplet.CName))
            {
                result.Rule = "1";
                result.ConcatenationPower = 0;
                result.TargetPremutation = triplet;
            }

            return result;
        }
    }

    internal class JoinBetweenTiers2 : IJoinMethod
    {
        private static bool Follows(int aIndex, int bIndex)
        {
            return aIndex == bIndex - 1;
        }

        public JoinInfo GetJoinInfo(ITabularFormula formula, ITriplet triplet)
        {
            var result = new JoinInfo(this);

            int aIndex = formula.Premutation.IndexOf(triplet.AName);
            int bIndex = formula.Premutation.IndexOf(triplet.BName);
            int cIndex = formula.Premutation.IndexOf(triplet.CName);
            
            if (Follows(aIndex, bIndex) && cIndex < 0)
            {
                var aTiers = formula.Tiers.Where(tier => tier[0].HasVariable(triplet.AName));
                var bTiers = formula.Tiers.Where(tier => tier[0].HasVariable(triplet.BName));

                var aTier = aTiers.FirstOrDefault();
                var bTier = bTiers.FirstOrDefault();

                if (aTiers.Count() == 1 && bTiers.Count() == 1 
                    && aTier != null && bTier != null && aTier != bTier)
                {
                    result.Rule = "1";
                    result.ConcatenationPower = 2;
                    result.TargetPremutation = new SimpleTriplet(triplet.AName, triplet.CName, triplet.BName);
                }
            }
            else if (Follows(cIndex, bIndex) && aIndex < 0)
            {
                var cTiers = formula.Tiers.Where(tier => tier[0].HasVariable(triplet.CName));
                var bTiers = formula.Tiers.Where(tier => tier[0].HasVariable(triplet.BName));

                var cTier = cTiers.FirstOrDefault();
                var bTier = bTiers.FirstOrDefault();

                if (cTiers.Count() == 1 && bTiers.Count() == 1
                    && cTier != null && bTier != null && cTier != bTier)
                {
                    result.Rule = "2";
                    result.ConcatenationPower = 2;
                    result.TargetPremutation = new SimpleTriplet(triplet.CName, triplet.AName, triplet.BName);
                }
            }
            else if (Follows(aIndex, cIndex) && bIndex < 0)
            {
                var aTiers = formula.Tiers.Where(tier => tier[0].HasVariable(triplet.AName));
                var cTiers = formula.Tiers.Where(tier => tier[0].HasVariable(triplet.CName));

                var aTier = aTiers.FirstOrDefault();
                var cTier = cTiers.FirstOrDefault();

                if (aTiers.Count() == 1 && cTiers.Count() == 1
                    && aTier != null && cTier != null && aTier != cTier)
                {
                    result.Rule = "3";
                    result.ConcatenationPower = 2;
                    result.TargetPremutation = new SimpleTriplet(triplet.AName, triplet.BName, triplet.CName);
                }
            }

            return result;
        }
    }
}

File: java/3-sat-core/src/main/java/com/anjlab/sat3/JoinMethods.java
package com.anjlab.sat3;


public class JoinMethods
{
	private static final IJoinMethod[] joinMethods = 
						new IJoinMethod[]
		                {
							//	The order matters (see comments in Helper#getJoinCandidates() and Helper#pickAJoin()
							new Join3AsIs(),
					        new JoinBetweenTiers2(),
					        new JoinLeft2(),
					        new JoinRight2(),
					        new JoinLeft1(),
					        new JoinRight1(),
					        new JoinRight0(),
					    };  
	
    public static IJoinMethod[] getMethods()
    {
        return joinMethods;
    }
}

class Join3AsIs implements IJoinMethod
{
    public JoinInfo getJoinInfo(ITabularFormula formula, ITier tier)
    {
        JoinInfo result = new JoinInfo(this);

        int aIndex = formula.getPermutation().indexOf(tier.getAName());
        int bIndex = formula.getPermutation().indexOf(tier.getBName());
        int cIndex = formula.getPermutation().indexOf(tier.getCName());

        if (aIndex >= 0 && bIndex >= 0 && cIndex >= 0)
        {
            if (follows(aIndex, bIndex, cIndex))
            {
                result.rule = "1";
                result.concatenationPower = 3;
                result.targetPermutation = tier;
            }
            else if (follows(bIndex, aIndex, cIndex))
            {
                result.rule = "2";
                result.concatenationPower = 3;
                result.targetPermutation = new SimpleTripletPermutation(tier.getBName(), tier.getAName(), tier.getCName());
            }
            else if (follows(aIndex, cIndex, bIndex))
            {
                result.rule = "3";
                result.concatenationPower = 3;
                result.targetPermutation = new SimpleTripletPermutation(tier.getAName(), tier.getCName(), tier.getBName());
            }
            else if (follows(cIndex, aIndex, bIndex))
            {
                result.rule = "4";
                result.concatenationPower = 3;
                result.targetPermutation = new SimpleTripletPermutation(tier.getCName(), tier.getAName(), tier.getBName());
            }
            else if (follows(bIndex, cIndex, aIndex))
            {
                result.rule = "5";
                result.concatenationPower = 3;
                result.targetPermutation = new SimpleTripletPermutation(tier.getBName(), tier.getCName(), tier.getAName());
            }
            else if (follows(cIndex, bIndex, aIndex))
            {
                result.rule = "6";
                result.concatenationPower = 3;
                result.targetPermutation = new SimpleTripletPermutation(tier.getCName(), tier.getBName(), tier.getAName());
            }
        }

        return result;
    }

    private static boolean follows(int aIndex, int bIndex, int cIndex)
    {
        return follows(aIndex, bIndex) && follows(bIndex, cIndex);
    }

    private static boolean follows(int aIndex, int bIndex)
    {
        return aIndex == bIndex - 1;
    }
}

class JoinLeft1 implements IJoinMethod
{
    public JoinInfo getJoinInfo(ITabularFormula formula, ITier tier)
    {
        JoinInfo result = new JoinInfo(this);

        if (formula.getPermutation().get(0) == tier.getCName()
            && !formula.getPermutation().contains(tier.getBName())
            && !formula.getPermutation().contains(tier.getAName()))
        {
            result.rule = "1";
            result.concatenationPower = 1;
            result.targetPermutation = new SimpleTripletPermutation(tier.getAName(), tier.getBName(), tier.getCName());
        }
        else if (formula.getPermutation().get(0) == tier.getBName()
            && !formula.getPermutation().contains(tier.getAName())
            && !formula.getPermutation().contains(tier.getCName()))
        {
            result.rule = "2";
            result.concatenationPower = 1;
            result.targetPermutation = new SimpleTripletPermutation(tier.getAName(), tier.getCName(), tier.getBName());
        }
        else if (formula.getPermutation().get(0) == tier.getAName()
            && !formula.getPermutation().contains(tier.getBName())
            && !formula.getPermutation().contains(tier.getCName()))
        {
            result.rule = "3";
            result.concatenationPower = 1;
            result.targetPermutation = new SimpleTripletPermutation(tier.getBName(), tier.getCName(), tier.getAName());
        }

        return result;
    }
}

class JoinLeft2 implements IJoinMethod
{
    public JoinInfo getJoinInfo(ITabularFormula formula, ITier tier)
    {
        JoinInfo result = new JoinInfo(this);

        if (formula.getPermutation().get(0) == tier.getBName()
            && formula.getPermutation().get(1) == tier.getCName()
            && !formula.getPermutation().contains(tier.getAName()))
        {
            result.rule = "1";
            result.concatenationPower = 2;
            result.targetPermutation = new SimpleTripletPermutation(tier.getAName(), tier.getBName(), tier.getCName());
        }
        else if (formula.getPermutation().get(0) == tier.getCName()
            && formula.getPermutation().get(1) == tier.getBName()
            && !formula.getPermutation().contains(tier.getAName()))
        {
            result.rule = "2";
            result.concatenationPower = 2;
            result.targetPermutation = new SimpleTripletPermutation(tier.getAName(), tier.getCName(), tier.getBName());
        }
        else if (formula.getPermutation().get(0) == tier.getAName()
            && formula.getPermutation().get(1) == tier.getBName()
            && !formula.getPermutation().contains(tier.getCName()))
        {
            result.rule = "3";
            result.concatenationPower = 2;
            result.targetPermutation = new SimpleTripletPermutation(tier.getCName(), tier.getAName(), tier.getBName());
        }
        else if (formula.getPermutation().get(0) == tier.getBName()
            && formula.getPermutation().get(1) == tier.getAName()
            && !formula.getPermutation().contains(tier.getCName()))
        {
            result.rule = "4";
            result.concatenationPower = 2;
            result.targetPermutation = new SimpleTripletPermutation(tier.getCName(), tier.getBName(), tier.getAName());
        }
        else if (formula.getPermutation().get(0) == tier.getAName()
            && formula.getPermutation().get(1) == tier.getCName()
            && !formula.getPermutation().contains(tier.getBName()))
        {
            result.rule = "5";
            result.concatenationPower = 2;
            result.targetPermutation = new SimpleTripletPermutation(tier.getBName(), tier.getAName(), tier.getCName());
        }
        else if (formula.getPermutation().get(0) == tier.getCName()
            && formula.getPermutation().get(1) == tier.getAName()
            && !formula.getPermutation().contains(tier.getBName()))
        {
            result.rule = "6";
            result.concatenationPower = 2;
            result.targetPermutation = new SimpleTripletPermutation(tier.getBName(), tier.getCName(), tier.getAName());
        }

        return result;
    }
}

class JoinRight2 implements IJoinMethod
{
    public JoinInfo getJoinInfo(ITabularFormula formula, ITier tier)
    {
        JoinInfo result = new JoinInfo(this);

        if (formula.getPermutation().get(formula.getPermutation().size() - 2) == tier.getAName()
            && formula.getPermutation().get(formula.getPermutation().size() - 1) == tier.getBName()
            && !formula.getPermutation().contains(tier.getCName()))
        {
            result.rule = "1";
            result.concatenationPower = 2;
            result.targetPermutation = tier;
        }
        else if (formula.getPermutation().get(formula.getPermutation().size() - 2) == tier.getBName()
            && formula.getPermutation().get(formula.getPermutation().size() - 1) == tier.getAName()
            && !formula.getPermutation().contains(tier.getCName()))
        {
            result.rule = "2";
            result.concatenationPower = 2;
            result.targetPermutation = new SimpleTripletPermutation(tier.getBName(), tier.getAName(), tier.getCName());
        }
        else if (formula.getPermutation().get(formula.getPermutation().size() - 2) == tier.getBName()
            && formula.getPermutation().get(formula.getPermutation().size() - 1) == tier.getCName()
            && !formula.getPermutation().contains(tier.getAName()))
        {
            result.rule = "3";
            result.concatenationPower = 2;
            result.targetPermutation = new SimpleTripletPermutation(tier.getBName(), tier.getCName(), tier.getAName());
        }
        else if (formula.getPermutation().get(formula.getPermutation().size() - 2) == tier.getCName()
            && formula.getPermutation().get(formula.getPermutation().size() - 1) == tier.getBName()
            && !formula.getPermutation().contains(tier.getAName()))
        {
            result.rule = "4";
            result.concatenationPower = 2;
            result.targetPermutation = new SimpleTripletPermutation(tier.getCName(), tier.getBName(), tier.getAName());
        }
        else if (formula.getPermutation().get(formula.getPermutation().size() - 2) == tier.getAName()
            && formula.getPermutation().get(formula.getPermutation().size() - 1) == tier.getCName()
            && !formula.getPermutation().contains(tier.getBName()))
        {
            result.rule = "5";
            result.concatenationPower = 2;
            result.targetPermutation = new SimpleTripletPermutation(tier.getAName(), tier.getCName(), tier.getBName());
        }
        else if (formula.getPermutation().get(formula.getPermutation().size() - 2) == tier.getCName()
            && formula.getPermutation().get(formula.getPermutation().size() - 1) == tier.getAName()
            && !formula.getPermutation().contains(tier.getBName()))
        {
            result.rule = "6";
            result.concatenationPower = 2;
            result.targetPermutation = new SimpleTripletPermutation(tier.getCName(), tier.getAName(), tier.getBName());
        }

        return result;
    }
}

class JoinRight1 implements IJoinMethod
{
    public JoinInfo getJoinInfo(ITabularFormula formula, ITier tier)
    {
        JoinInfo result = new JoinInfo(this);

        if (formula.getPermutation().get(formula.getPermutation().size() - 1) == tier.getAName()
            && !formula.getPermutation().contains(tier.getBName())
            && !formula.getPermutation().contains(tier.getCName()))
        {
            result.rule = "1";
            result.concatenationPower = 1;
            result.targetPermutation = tier;
        }
        else if (formula.getPermutation().get(formula.getPermutation().size() - 1) == tier.getBName()
            && !formula.getPermutation().contains(tier.getAName())
            && !formula.getPermutation().contains(tier.getCName()))
        {
            result.rule = "2";
            result.concatenationPower = 1;
            result.targetPermutation = new SimpleTripletPermutation(tier.getBName(), tier.getAName(), tier.getCName());
        }
        else if (formula.getPermutation().get(formula.getPermutation().size() - 1) == tier.getCName()
            && !formula.getPermutation().contains(tier.getAName())
            && !formula.getPermutation().contains(tier.getBName()))
        {
            result.rule = "3";
            result.concatenationPower = 1;
            result.targetPermutation = new SimpleTripletPermutation(tier.getCName(), tier.getAName(), tier.getBName());
        }

        return result;
    }
}

class JoinRight0 implements IJoinMethod
{
    public JoinInfo getJoinInfo(ITabularFormula formula, ITier tier)
    {
        JoinInfo result = new JoinInfo(this);

        if (!formula.getPermutation().contains(tier.getAName())
            && !formula.getPermutation().contains(tier.getBName())
            && !formula.getPermutation().contains(tier.getCName()))
        {
            result.rule = "1";
            result.concatenationPower = 0;
            result.targetPermutation = tier;
        }

        return result;
    }
}

class JoinBetweenTiers2 implements IJoinMethod
{
    private static boolean follows(int aIndex, int bIndex)
    {
        return aIndex == bIndex - 1;
    }

    public JoinInfo getJoinInfo(ITabularFormula formula, ITier tier)
    {
        JoinInfo result = new JoinInfo(this);

        int aIndex = formula.getPermutation().indexOf(tier.getAName());
        int bIndex = formula.getPermutation().indexOf(tier.getBName());
        int cIndex = formula.getPermutation().indexOf(tier.getCName());
        
        if (follows(aIndex, bIndex) && cIndex < 0)
        {
        	GenericArrayList<ITier> aTiers = findTiersWithVariable(formula.getTiers(), tier.getAName());
        	GenericArrayList<ITier> bTiers = findTiersWithVariable(formula.getTiers(), tier.getBName());

            ITier aTier = aTiers.size() > 0 ? aTiers.get(0) : null;
            ITier bTier = bTiers.size() > 0 ? bTiers.get(0) : null;

            if (aTiers.size() == 1 && bTiers.size() == 1 
                && aTier != null && bTier != null && aTier != bTier)
            {
                result.rule = "1";
                result.concatenationPower = 2;
                result.targetPermutation = new SimpleTripletPermutation(tier.getAName(), tier.getCName(), tier.getBName());
            }
        }
        else if (follows(cIndex, bIndex) && aIndex < 0)
        {
        	GenericArrayList<ITier> cTiers = findTiersWithVariable(formula.getTiers(), tier.getCName());
        	GenericArrayList<ITier> bTiers = findTiersWithVariable(formula.getTiers(), tier.getBName());

            ITier cTier = cTiers.size() > 0 ? cTiers.get(0) : null;
            ITier bTier = bTiers.size() > 0 ? bTiers.get(0) : null;

            if (cTiers.size() == 1 && bTiers.size() == 1
                && cTier != null && bTier != null && cTier != bTier)
            {
                result.rule = "2";
                result.concatenationPower = 2;
                result.targetPermutation = new SimpleTripletPermutation(tier.getCName(), tier.getAName(), tier.getBName());
            }
        }
        else if (follows(aIndex, cIndex) && bIndex < 0)
        {
            GenericArrayList<ITier> aTiers = findTiersWithVariable(formula.getTiers(), tier.getAName());
            GenericArrayList<ITier> cTiers = findTiersWithVariable(formula.getTiers(), tier.getCName());

            ITier aTier = aTiers.size() > 0 ? aTiers.get(0) : null;
            ITier cTier = cTiers.size() > 0 ? cTiers.get(0) : null;

            if (aTiers.size() == 1 && cTiers.size() == 1
                && aTier != null && cTier != null && aTier != cTier)
            {
                result.rule = "3";
                result.concatenationPower = 2;
                result.targetPermutation = new SimpleTripletPermutation(tier.getAName(), tier.getBName(), tier.getCName());
            }
        }

        return result;
    }

	private GenericArrayList<ITier> findTiersWithVariable(GenericArrayList<ITier> tiers, int varName) {
		GenericArrayList<ITier> result = new GenericArrayList<ITier>();
		for (int i = 0; i < tiers.size(); i++)
		{
			ITier tier = tiers.get(i);
			if (tier.hasVariable(varName))
			{
				result.add(tier);
			}
		}
		return result;
	}
}